### ISSUE TITLE
NetBSD support

### ISSUE THREAD

**Issue by krytarowski** [frida/frida#429]
I'm trying to install Frida through `r2pm -i r2frida`

I face the following problem:

```
curl -Ls https://github.com/frida/frida/releases/download/10.6.41/frida-core-devkit-10.6.41-netbsd-amd64.tar.xz | xz -d | tar -C ext/frida-netbsd-10.6.41/_ -xf -
xz: (stdin): File format not recognized
```

The file does not exist.

**Comment by enovella** [comment]
Hi @krytarowski,

Can you see where the problem is by doing command to command? For instance, 

1. Do `curl`
2. Do `curl` + xz
3. Do `curl` + xz + tar
4. ....

You will find the issue by yourself.

Cheers

**Comment by oleavr** [comment]
Hey @krytarowski,

The issue is that Frida does not yet support NetBSD. We would love to support it, though, so PRs are most welcome! You can find some pointers on porting Frida to new platforms at: https://www.frida.re/docs/hacking/

**Comment by krytarowski** [comment]
I will have a look. Hope it's fine to have a subset of backends (`frida-*`) for now.

**Comment by oleavr** [comment]
@krytarowski Awesome! Yeah it's fine to start out with most things being stubs, e.g. the `foo_enumerate_bar` APIs could just be implemented as empty functions that don't call the callback, effectively returning zero results.

**Comment by krytarowski** [comment]
I'm using manually `build-targetfunctions.sh` to generate targetfunctions and special functions for NetBSD/amd64:

```
CFLAGS="-fPIC" STRIP=strip CC=gcc CONFIG_SITE=$PWD PKG_CONFIG=pkg-config LDFLAGS="-fPIC -DPIC" sh build-targetfunctions.sh netbsd x86_64
```

Is this file supposed to be called manually in such way? It could use libtool for portability reasons.

Another question is regarding glib and libffi, it seems I need to use both from forks from frida. Anything else custom needed?

**Comment by oleavr** [comment]
(Sorry for the delay here, I was knocked out by a little food-poisoning incident over the last few days.)

> Is this file supposed to be called manually in such way? It could use libtool for portability reasons.

No it should be run from inside a Frida build environment, which is generated by [setup-env.sh](https://github.com/frida/frida/blob/master/releng/setup-env.sh). That script should only hopefully only need one `netbsd` block similar to the one for [linux](https://github.com/frida/frida/blob/72d6333e5ac6d5d50702e656a3d2671863ee848c/releng/setup-env.sh#L185-L250). Once that's in place, and `Makefile.sdk.mk` is able to build an SDK – `make -f Makefile.sdk.mk FRIDA_HOST=netbsd-x86_64` – run `build-targetfunctions.sh` and git commit the resulting binaries. (They're intended to not change so that Interceptor sees the exact same instruction sequences over time, i.e. the functions being instrumented don't change.)

> Another question is regarding glib and libffi, it seems I need to use both from forks from frida.

Correct.

> Anything else custom needed?

Not that I can remember – the dependencies built by `Makefile.sdk.mk` should be everything needed for a given platform beside compiler/toolchain/libc headers. The reason we maintain our own forks of GLib and libffi is that we need to be able to clean up statically allocated resources, as `frida-agent.so` may be injected into long-running processes, making it desirable to unload it at some point. (And be able to load it again without leaking resources.)

**Comment by krytarowski** [comment]
> (Sorry for the delay here, I was knocked out by a little food-poisoning incident over the last few days.)

I hope everything will be fine!

>>    Another question is regarding glib and libffi, it seems I need to use both from forks from frida.
> Correct.

OK, I'm going to package them and attempt to build frida-gum.

**Comment by oleavr** [comment]
> I hope everything will be fine!

Thanks! Back on my feet today 😊

> OK, I'm going to package them and attempt to build frida-gum.

Awesome! Don't hesitate to poke me, nick: `oleavr`, in the [fridadotre channel](https://t.me/fridadotre) on Telegram or its mirror on IRC (#frida on FreeNode) whenever something's unclear.

**Comment by krytarowski** [comment]
Is there some way to test frida-gum with ported libffi, glib? Do I need core and maybe python to get anything to work? I should get them usable in not so distant time.

**Comment by oleavr** [comment]
Awesome!

The `gum-tests` binary (built by the build system) should provide pretty good coverage. But maybe a better “Hello World” would be a function hooking example. For that grab any one of the `frida-gum-devkit-*` tarballs from Frida’s [releases](https://github.com/frida/frida/releases) and compile the example.c inside of it against your Gum — something like: `$CC example.c $(pkg-config —cflags —libs frida-gum-1.0)`

**Comment by krytarowski** [comment]
It looks I need frida-capstone too.

```
../gum/arch-x86/gumx86relocator.c:555:12: error: no member named 'modrm_offset' in 'struct cs_x86'
  if (x86->modrm_offset == 0)
      ~~~  ^
../gum/arch-x86/gumx86relocator.c:617:15: error: no member named 'modrm_offset' in 'struct cs_x86'
    code[x86->modrm_offset] = (mod << 6) | (reg << 3) | rm;
         ~~~  ^
2 errors generated.
```

**Comment by oleavr** [comment]
> It looks I need frida-capstone too.

That’s unfortunately correct.

**Comment by krytarowski** [comment]
Hmm, I've been a little bit surprised that OS specific bits are larger than expected. Once I will build frida-gum with stubs, I will start upstreaming them in chunks.

While there - this is how I landed here in frida... I'm researching a room for a new self-introspecting syscall in the NetBSD kernel. More context: https://github.com/ivmai/bdwgc/issues/207

Frida needs the same horrid magic on Linux: https://lkml.org/lkml/2006/9/1/217

I'm looking for a feedback whether it makes sense, and maybe how to extend it with additional features. I'm thinking to add 2 operations useful for remote forking: serializing and deserializing an application (something close to core-dumps). I would want it for criu, rr and reptyr.

Are there wanted features from the Frida point of view that could make life easier?

**Comment by oleavr** [comment]
> Hmm, I've been a little bit surprised that OS specific bits are larger than expected.

Indeed. The good news is that once Gum has been ported there is only one part left needing porting: the shared library injector in frida-core. Since NetBSD supports `ptrace()` it might be possible to take inspiration from the Linux implementation. (It's pretty horrible though as it's doing thread hi-jacking trickery due to lack of a better API than `ptrace()` – our i/macOS and Windows backends are simpler in that regard due to more suitable OS APIs being available.)

> Once I will build frida-gum with stubs, I will start upstreaming them in chunks.

Sounds great!

> I'm researching a room for a new self-introspecting syscall in the NetBSD kernel

Cool!

> Frida needs the same horrid magic on Linux: https://lkml.org/lkml/2006/9/1/217

Yep, that's how we implement [Gum.Process.modify_thread()](https://github.com/frida/frida-gum/blob/c41abae2a2b29de3aad7a4fd9daac3e0b150cb54/gum/backend-linux/gumprocess-linux.c#L384-L410). It's horrible.

> Are there wanted features from the Frida point of view that could make life easier?

I'm not familiar with what's currently available on NetBSD, but ideally there would be support for process A asking the kernel to perform the following operations in the context of process B:
- Allocate pages of memory.
- (Read/)Write (from/)to memory.
- Change page protections.
- Spawn a new thread, fully initialized so libc can be used.

On Linux we currently emulate these things by hi-jacking an existing thread with `ptrace()`.

**Comment by krytarowski** [comment]
I work on modernization of debugging interfaces and tools:

http://netbsd.org/~kamil/ptrace-netbsd/presentation.html (bit dates, but gives a good overview)
http://netbsd.org/~kamil/eurobsdcon2017.html (LLDB / NetBSD)

The general difference between Linux and NetBSD is that ptrace(2) is per-process (process group on Linux) rather than per-thread (process on Linux).

Frida is definitely a modern (and innovative) tool that we want to get aboard.

> Yep, that's how we implement Gum.Process.modify_thread(). It's horrible.

Right. As of now we don't have a support for this, at least in a relatively clean approach. It could be reimplemented with fork(2) + ptrace(2) and IPC; or signal handlers.. but all that is horrid, racy and dangerous.

If this is a hard-requirement of getting Frida aboard, I will port the remaining bits, upstream, and keep researching a kernel new syscall followed by an implementation.

> I'm not familiar with what's currently available on NetBSD, but ideally there would be support for process A asking the kernel to perform the following operations in the context of process B:
>
>    Allocate pages of memory.
>    (Read/)Write (from/)to memory.
>    Change page protections.
>    Spawn a new thread, fully initialized so libc can be used.
>
> On Linux we currently emulate these things by hi-jacking an existing thread with ptrace().

Perfect! This is what I was looking for.

LLDB people told me that it would nice to have a remote ability to execute a syscall in the contest of traced process/thread... but I don't know of use-cases.

Regarding thread.. I expect that we don't need a pthread entity? That would be harder than a native thread (LWP in NetBSD), as it would involve changes not just in the kernel, but also in libpthread structures.

I think once we can get a new thread and control its body, we can do everything from it, including memory operations and the focus shall be on starting of a new injected thread.

Low-level code for starting of a thread in userspace looks like:

```
_lwp_makecontext(&uc, lwp_main_func, NULL, NULL, stack, ssize);
_lwp_create(&uc, 0, &lid);
```

This means that we shall be already in the context of a target process and have dedicated text and data/bss piece of section.

Assuming that we can fork LWP (thread) from the kernel, we would need to inject new code (stack, ...) to execute it.

Unfortunately it looks complex and enrolls some assumptions and it will be followed by portability issues like static vs shared program code.

Maybe I need to introduce a trampoline for `pinspect` thread in libc, created remotely from `ptrace`..

I'm thinking out loudly looking for a solution.

**Comment by oleavr** [comment]
> I work on modernization of debugging interfaces and tools

Awesome!!

> Frida is definitely a modern (and innovative) tool that we want to get aboard.

Cool!! Glad to hear that!

> If this is a hard-requirement of getting Frida aboard

The primary use-case is [Stalker](https://github.com/frida/frida-presentations/blob/master/R2Con2017/02-transforms/04-aes.js), which uses it to dynamically instrument the threads of your choosing. We also piggyback on it for enumerating threads. But a lot of users won't care about either of these two things, so it's functionality that could be implemented down the road – Frida is also useful without that functionality.

> Regarding thread.. I expect that we don't need a pthread entity?

Well `pthread_self()` should work and the thread needs to be able to spawn new pthreads. Though if `pthread_create()` works from an LWP we could just do that in the bootstrapper that we inject, to create a "full thread" and carry on from there.

Apple solved this on newer OSes by introducing an undocumented version of `pthread_create()` named [pthread_create_from_mach_thread()](https://github.com/frida/frida-core/blob/42c026301fb837ab3757c555a65e479804a9344c/src/darwin/frida-helper-backend-glue.m#L2238), which we use when available. It is essentially a "safer" version of `pthread_create()` for use in naked Mach threads. That thread is still not able to use any other pthread APIs though, so we block in a [mach_msg_receive()](https://github.com/frida/frida-core/blob/42c026301fb837ab3757c555a65e479804a9344c/src/darwin/frida-helper-backend-glue.m#L2398-L2399) forever, making the thread safe to terminate. (It's horrible, but it's what Apple themselves do in Xcode's shared library injection implementation.)

So yeah perhaps the easiest for now is just using the same hijacking mechanism we use on Linux?
 That is assuming that it's always safe to interrupt a system call and later resume it. There's also the risk of hijacking a thread while it's inside pthread holding some lock so `pthread_create()` isn't safe... but I guess we can live with that, at least initially.

Though OTOH if you have a clear idea how to add the needed kernel support nothing would be better... it just sounds like a lot of added work on top of the other porting TODOs. :-)

**Comment by krytarowski** [comment]
> Though OTOH if you have a clear idea how to add the needed kernel support nothing would be better...

"How" is just a matter of time an digging, I focus on the goal. I'm researching use-cases/users that will shape the API, in the context of what is possible.

Regarding the Linux approach, it's not portable as-is. NetBSD's solution would be more complicated and more horrid... and I have got enough abomination before starting it..

Perhaps something shaped close to signal handler/trampoline that executes code in libc is the way to go, establishing IPC between parent and child, and calling pinspect().. I need to think more about it. I think that the hard-requirement to make it Machine-Independent.

Thank you for the list of requirements.

>  it just sounds like a lot of added work on top of the other porting TODOs. :-)

My porting is driven by this.

**Comment by oleavr** [comment]
Very cool!! Looking forward to see how it evolves!

**Comment by krytarowski** [comment]
It looks like the only reliable way to perform anything from process B in process A.. is to rewrite .text/registers in process A using ptrace(2) from process B.

I'm in the process of fixing kernel bugs in NetBSD, mostly with multiple threads, signals. forks/vforks. This is planned to take a lot of time.. I will follow up by StopTheWorld-like syscall (from the current-process context).

Are there plans to upstream glib2, libffi and capstone patches for Frida? It would make life easier to use pristine versions... while there I feel motivated to streamline NetBSD in glib2, as we keep over 30 local patches in pkgsrc... Frida uses old forks and I need to at least research whether there is need to backport them.

**Comment by oleavr** [comment]
> It looks like the only reliable way to perform anything from process B in process A.. is to rewrite .text/registers in process A using ptrace(2) from process B.

OK!

> Are there plans to upstream glib2, libffi and capstone patches for Frida?

- GLib: No. Beside the bug-fixes, which are upstreamed now and then, we have some custom APIs that won't ever make it upstream. The big one here is that they don't consider it a valid use-case to be able to clean up static allocations/support unloading. Also there isn't that much value in using a system-wide GLib, as we always want to link it statically to allow users to instrument applications using the system-wide GLib. (Due to reentrancy-issues when hooking low-level GLib APIs.) Btw, I just rebased our GLib patches on top of the latest master the other day – it's living in the `next` branch and should soon become `master`.
- libffi: I've upstreamed everything except `Add API for custom memory management`, which I'm quite sure that upstream won't be interested in as it's for a really specialized use-case: knowing about all of libffi's memory allocations so we can cloak them -- pretend they don't exist when the user does `Process.enumerateRanges()`. There's also the need to clean up static allocations here as well, which is unlikely to fly. (But might be worth a shot.)
- capstone: Yes, but it's going to take a while. (It's way down on my list, spread a bit too thin these days.)

**Comment by krytarowski** [comment]
Looks good. Is there an option to namespace these libraries, like adding prefix frida- to libraries, pc, directory with headers and wherever needed?

I've packaged them in pkgsrc, but there are some annoying conflicts with mainstream packages.. and I think I will need to do it myself with local patches.

**Comment by krytarowski** [comment]
This part stops me with building everything of frida-gum. I will namespace files installed by glib2 etc.

**Comment by oleavr** [comment]
> Is there an option to namespace these libraries

No but it would be good to namespace the files. I'm not sure I would recommend installing them on the system though – only static versions should be built, which is required for the `deinit()` APIs to not be no-ops.
